File: src/main/java/org/moinex/model/InvestmentPerformanceSnapshot.java
/*
 * Filename: InvestmentPerformanceSnapshot.java
 * Created on: February 17, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "investment_performance_snapshot")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class InvestmentPerformanceSnapshot {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    @Setter(AccessLevel.NONE)
    private Integer id;

    @Column(name = "month", nullable = false)
    private Integer month;

    @Column(name = "year", nullable = false)
    private Integer year;

    @Column(name = "invested_value", nullable = false)
    private BigDecimal investedValue;

    @Column(name = "portfolio_value", nullable = false)
    private BigDecimal portfolioValue;

    @Column(name = "accumulated_capital_gains", nullable = false)
    private BigDecimal accumulatedCapitalGains;

    @Column(name = "monthly_capital_gains", nullable = false)
    private BigDecimal monthlyCapitalGains;

    @Column(name = "calculated_at", nullable = false)
    private String calculatedAt;
}

File: src/main/kotlin/org/moinex/model/InvestmentPerformanceSnapshot.kt
/*
 * Filename: InvestmentPerformanceSnapshot.kt (original filename: InvestmentPerformanceSnapshot.java)
 * Created on: February 17, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 07/03/2026
 */

package org.moinex.model

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import org.moinex.common.extension.toRounded
import java.math.BigDecimal

@Entity
@Table(name = "investment_performance_snapshot")
class InvestmentPerformanceSnapshot(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    var id: Int? = null,
    @Column(name = "month", nullable = false)
    var month: Int,
    @Column(name = "year", nullable = false)
    var year: Int,
    @Column(name = "invested_value", nullable = false)
    var investedValue: BigDecimal,
    @Column(name = "portfolio_value", nullable = false)
    var portfolioValue: BigDecimal,
    @Column(name = "accumulated_capital_gains", nullable = false)
    var accumulatedCapitalGains: BigDecimal,
    @Column(name = "monthly_capital_gains", nullable = false)
    var monthlyCapitalGains: BigDecimal,
    @Column(name = "calculated_at", nullable = false)
    var calculatedAt: String,
) {
    init {
        investedValue = investedValue.toRounded()
        portfolioValue = portfolioValue.toRounded()
        accumulatedCapitalGains = accumulatedCapitalGains.toRounded()
        monthlyCapitalGains = monthlyCapitalGains.toRounded()

        require(month in 1..12) {
            "Month must be between 1 and 12"
        }
        require(year > 0) {
            "Year must be positive"
        }
    }

    override fun toString(): String = "InvestmentPerformanceSnapshot [month=$month, year=$year]"
}
